# Homework 1

In [ ]:
__author__ = "Christopher Potts"
__version__ = "CS224u, Stanford, Spring 2018 term"

In [1]:
__student__ = "Sherry Ruan"
__ID__ = "06033382"

This homework covers material from the unit on distributed representations. The primary goal is to explore some new techniques for building and assessing VSMs. The code you write as part of the assignment should be useful for research involving vector representations as well.

Like all homeworks, this should be submitted via Canvas. All you have to do is paste in your answers (which are all numerical values) and include the SUNetIds of anyone you worked with. Here's a direct link to the homework form:

https://canvas.stanford.edu/courses/83399/quizzes/50268

__Contents__

0. [Questions 1–2: Dice distance [2 points]](#Questions-1–2:-Dice-distance-[2-points])
0. [Question 3: t-test reweighting [2 points]](#Question-3:-t-test-reweighting-[2-points])
0. [Questions 4–6: Reweighting and co-occurrence frequency [3 points]](#Questions-4–6:-Reweighting-and-co-occurrence-frequency-[3-points])
0. [Question 7: Meeting the GloVe objective [1 point]](#Question-7:-Meeting-the-GloVe-objective-[1-point])
0. [Question 8: Expressive eloooongation [2 points]](#Question-8:-Expressive-eloooongation-[2-points])

In [2]:
import numpy as np
import os
import pandas as pd
from mittens import GloVe
from scipy.stats import pearsonr
import vsm

## Questions 1–2: Dice distance [2 points]

First, implement [Dice distance](https://en.wikipedia.org/wiki/Sørensen–Dice_coefficient) for real-valued vectors of dimension $n$, as

$$\textbf{dice}(u, v) = 1 - \frac{
    2 \sum_{i=1}^{n}\min(u_{i}, v_{i})
}{
    \sum_{i=1}^{n} u_{i} + v_{i}
}$$

(You can use `vsm.matching` for part of this.)

In [6]:
def dice_distance(u, v):
    numerator = 2 * vsm.matching(u, v)
    denominator = np.sum(u+v)
    distance = 1 - numerator / denominator
    return distance

Second, you might want to test your implementation. Here's a simple function for that:

In [7]:
def test_dice_implementation(func):
    """`func` should be an implementation of `dice` as defined above."""
    X = np.array([
        [  4.,   4.,   2.,   0.],
        [  4.,  61.,   8.,  18.],
        [  2.,   8.,  10.,   0.],
        [  0.,  18.,   0.,   5.]]) 
    assert func(X[0], X[1]).round(5) == 0.80198
    assert func(X[1], X[2]).round(5) == 0.67568

In [8]:
test_dice_implementation(dice_distance)

Third, use your implementation to measure the distance between A and B and between B and C in the toy `ABC` matrix we used in the first VSM notebook, repeated here for convenience.

In [9]:
ABC = pd.DataFrame([
    [ 2.0,  4.0], 
    [10.0, 15.0], 
    [14.0, 10.0]],
    index=['A', 'B', 'C'],
    columns=['x', 'y']) 

ABC

x     y
A   2.0   4.0
B  10.0  15.0
C  14.0  10.0

In [18]:
ABC.loc['A']

x    2.0
y    4.0
Name: A, dtype: float64

In [23]:
dice_distance(ABC.loc['A'], ABC.loc['B'])

0.6129032258064516

In [24]:
dice_distance([2,4], [10,15])

0.6129032258064516

In [25]:
dice_distance(ABC.loc['B'], ABC.loc['C'])

0.18367346938775508

__To submit:__

1. Dice distance between A and B. (Answer: 0.6129032258064516)
2. Dice distance between B and C. (Answer: 0.18367346938775508)

(The real question, which these values answer, is whether this measure place A and B close together relative to B and C – our goal for that example.) 

## Question 3: t-test reweighting [2 points]

The t-test statistic can be thought of as a reweighting scheme. For a count matrix $X$, row index $i$, and column index $j$:

$$\textbf{ttest}(X, i, j) = 
\frac{
    P(X, i, j) - \big(P(X, i, *)P(X, *, j)\big)
}{
\sqrt{(P(X, i, *)P(X, *, j))}
}$$

where $P(X, i, j)$ is $X_{ij}$ divided by the total values in $X$, $P(X, i, *)$ is the sum of the values in row $i$ of $X$ divided by the total values in $X$, and $P(X, *, j)$ is the sum of the values in column $j$ of $X$ divided by the total values in $X$.

First, implement this reweighting scheme.

In [78]:
import math

def ttest(df):
    # row sum
    row_sum = df.sum(axis=1)
    #print(row_sum.shape)
    # column sum
    col_sum = df.sum(axis=0)
    #print(col_sum.shape)
    # total sum
    total_sum = row_sum.sum()
    
    ttest_df = df.copy()
    
#     for row_idx, row in df.iterrows():
#         for col_idx, entry in enumerate(row):
#             product = row_sum[row_idx]*col_sum[col_idx]
#             ttest_df[row_idx][col_idx] = (entry - product/total_sum) / math.sqrt(product)
           
        
    product = np.outer(row_sum, col_sum) 
    #print(product.shape)
    ttest_df = (ttest_df - product/total_sum) / np.sqrt(product)
    
    return ttest_df

In [79]:
X = pd.DataFrame(np.array([
    [  4.,   4.,   2.,   0.],
    [  4.,  61.,   8.,  18.],
    [  2.,   8.,  10.,   0.],
    [  0.,  18.,   0.,   5.]]))
ttest(X)

(4,)
(4,)
(4, 4)


0         1         2         3
0  0.330556 -0.076889  0.043212 -0.105318
1 -0.076889  0.038385 -0.108737  0.075745
2  0.043212 -0.108737  0.361111 -0.148942
3 -0.105318  0.075745 -0.148942  0.057669

In [80]:
X

0     1     2     3
0  4.0   4.0   2.0   0.0
1  4.0  61.0   8.0  18.0
2  2.0   8.0  10.0   0.0
3  0.0  18.0   0.0   5.0

Second, test your implementation:

In [81]:
def test_ttest_implementation(func):
    """`func` should be an implementation of ttest reweighting as defined above."""
    X = pd.DataFrame(np.array([
        [  4.,   4.,   2.,   0.],
        [  4.,  61.,   8.,  18.],
        [  2.,   8.,  10.,   0.],
        [  0.,  18.,   0.,   5.]]))    
    actual = np.array([
        [ 0.33056, -0.07689,  0.04321, -0.10532],
        [-0.07689,  0.03839, -0.10874,  0.07574],
        [ 0.04321, -0.10874,  0.36111, -0.14894],
        [-0.10532,  0.07574, -0.14894,  0.05767]])    
    predicted = func(X)
    assert np.array_equal(predicted.round(5), actual)

In [82]:
test_ttest_implementation(ttest)

(4,)
(4,)
(4, 4)


Third, apply your implementation to the matrix stored in `imdb_window5-scaled.csv.gz`.

__To submit__: the cell value for the row labeled _superb_ and the column labeled _movie_.

(The goal here is really to obtain a working implementation of $\textbf{ttest}$. It could be an ingredient in a winning bake-off entry!)

In [83]:
data_home = 'vsmdata'

imdb5 = pd.read_csv(
    os.path.join(data_home, 'imdb_window5-scaled.csv.gz'), index_col=0) # shape is 5000 x 5000


In [84]:
imdb5_ttest = ttest(imdb5)

(5000,)
(5000,)
(5000, 5000)


In [88]:
imdb5_ttest.loc['superb']['movie']

-0.0008427614547882427

## Questions 4–6: Reweighting and co-occurrence frequency [3 points]

We've seen that raw count matrices encode a lot of frequency information. This is not necessarily all bad (stronger words like _superb_ will be rarer than weak ones like _good_ in part because of their more specialized semantics), but we do hope that our reweighting schemes will get us away from these relatively mundane associations. Thus, for any reweighting scheme, we should ask about its correlation with the raw co-occurrence counts.

Your task: using [scipy.stats.pearsonr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.pearsonr.html), calculate the Pearson correlation coefficient between the raw count values of `imdb5` as loaded in the previous question and the values obtained from applying PMI and Positive PMI to this matrix, and from reweighting each row by its length norm (as defined in the first noteboook for this unit; `vsm.length_norm`). Note: `X.values.ravel()` will give you the vector of values in the `pd.DataFrame` instance `X`.

In [94]:
imdb5_pmi = vsm.pmi(imdb5, positive=False)
pearsonr(imdb5.values.ravel(), imdb5_pmi.values.ravel())

(0.007653071529112714, 2.2613e-320)

In [95]:
imdb5_positive_pmi = vsm.pmi(imdb5)
pearsonr(imdb5.values.ravel(), imdb5_positive_pmi.values.ravel())

(0.04498063308409067, 0.0)

In [101]:
imdb5_normed = imdb5.apply(vsm.length_norm, axis=1)
pearsonr(imdb5.values.ravel(), imdb5_normed.values.ravel())

(0.12218662330676433, 0.0)

__To submit:__

1. Correlation coefficient for the PMI comparison.
1. Correlation coefficient for the Positive PMI comparison.
1. Correlation coefficient for the length-norm comparison.

(The hope is that seeing these values will give you a better sense for how these reweighting schemes compare to the input count matrices.)

## Question 7: Meeting the GloVe objective [1 point]

We saw that GloVe can be thought of as seeking vectors whose dot products are proportional to their PMI values. How close does GloVe come to this in practice? This question asks you to conduct a simple empirical assessment of that: 

1. Load the matrix stored as `imdb_window5-scaled.csv.gz` in the data distribution. Call this `imdb5`.
2. Reweight `imdb5` with Positive PMI.
3. Run GloVe on `imdb5` for 10 iterations, learning vectors of dimension 20 (`n=20`). Definitely use the implementation in the `mittens` package, not in `vsm.glove`, else this will take way too long. Except for `max_iter` and `n`, use all the default parameters.
4. Report the correlation between the cell values in the PMI and GloVe versions. For this, you can include all 0 values (even though GloVe ignores them). Use `pearsonr` as above.

In [104]:
glove_model = GloVe(n=20, max_iter=10)

imdb5_glv = glove_model.fit(imdb5.values)

imdb5_glv = pd.DataFrame(imdb5_glv, index=imdb5.index)

Iteration 10: loss: 1487575.75

In [115]:
M = imdb5_glv.dot(imdb5_glv.T)
print(M.shape)
np.corrcoef(M.values.ravel(), imdb5_positive_pmi.values.ravel()) # `np.corrcoef` actually already does `pearsonr`

(5000, 5000)


array([[1.        , 0.00299817],
       [0.00299817, 1.        ]])

In [109]:
pearsonr(M.values.ravel(), imdb5_positive_pmi.values.ravel())

(0.0029981636522754566, 8.426630791441401e-51)

## Question 8: Expressive eloooongation [2 points]

One of the goals of subword modeling is to capture out-of-vocabulary (OOV) words. This is particularly important for __expressive elogations__ like _coooooool_ and _booriiiing_. Because the amount of elongation is highly variable, we're unlikely to have good representations for such words. How does [our simple approach to subword modeling](vsm_01_distributional.ipynb#Subword-information) do with these phenomena?

__Your task:__

* Use `vsm.ngram_vsm` to create a 4-gram character-level VSM from the matrix in `imdb_window20-flat.csv.gz`.

* Using `character_level_rep` from the notebook for representing words in this space, calculate the cosine distances for pair `cool` and `cooooool`.

__To submit__: the cosine distance  between `cool` and `cooooool`

(Of course, the broader question we want to answer is whether these words are being modeled as similar, which is a more subjective, comparative question. It does depend on these distance calculations, though.)

In [117]:
imdb20 = pd.read_csv(
    os.path.join(data_home, 'imdb_window20-flat.csv.gz'), index_col=0)

imdb20.shape

(5000, 5000)

In [118]:
imdb20_ngrams = vsm.ngram_vsm(imdb20, n=4)

imdb20_ngrams.shape

(10167, 5000)

In [119]:
def character_level_rep(word, cf, n=4):
    ngrams = vsm.get_character_ngrams(word, n)
    ngrams = [n for n in ngrams if n in cf.index]    
    reps = cf.loc[ngrams].values
    return reps.sum(axis=0) 

In [121]:
cooooool = character_level_rep("cooooool", imdb20_ngrams)

In [122]:
cool = character_level_rep("cool", imdb20_ngrams)

In [123]:
vsm.cosine(cooooool, cool)

0.0006569654844812423